meteor/client/lib/editAttribute/edit-attribute-dropdown.tsx

```diff
@@ -1,161 +1,134 @@
 import { EditAttributeBase, IEditAttributeBaseProps, wrapEditAttribute } from './edit-attribute-base'
 import ClassNames from 'classnames'
 import * as React from 'react'
+import { WithTranslation, withTranslation } from 'react-i18next'
 
 export interface DropdownOption {
-	value: string
-	alternativeValue?: string
+	value: string | number
 	label?: string
 }
 
 interface EditAttributeDropdownProps extends IEditAttributeBaseProps {
 	options: DropdownOption[]
 	shouldSaveLabel?: boolean
+	allowNoSelection?: boolean
```

Comment:
We should try to strive for not having "boolean flags" as they are a bad practice/code smell (I am aware of the irony of me commenting this, since I made the `shoudSaveLabel`).

It looks like you are using `allowNoSelection` to check whether you should prepend an "empty" option to the options array.
A suggestion to get rid of the boolean flag would be to create another component that then utilizes `EditAttributeDropdown`.
It would look something like this:
```
export function EditAttributeDropdownEmptyFirstValue(props: EditAttributeDropdownProps) {
  const options: DropdownOption[] = prependEmptyOption(props.options)
  return <EditAttributeDropdown {...props} options={options}
}

function prependEmptyOption(options: DropdownOption[]): DropdownOption[] {
  return [{ value: '', label: t('None') }, ...this.props.options]
}
```
Then the places that needs the option to allow not having anything selected just uses the new component.

With the above code you have separated the responsibility of the "empty option" out of `EditAttributeDropdown` (which makes the implementation of `EditAttribueDropdown` simpler) and `EditAttributeDropdownEmptyFirstValue` only has one responsibility. And you got rid of a boolean flag :D

Reply:
I think your argument is invalid in regards to highly configurable UI components. We already have four boolean properties on `IEditAttributeBaseProps`, so a total of 6 in `EditAttributeDropdownProps`. Are you saying that we should strive to make 64 different component types for every combination of those _boolean flags_?

Reply:
No, I'm not saying we should make 64 variations. The fact that it has so many boolean flags is more a sign of bad design.

`EditAttributeBaseProps` is in my opinion a bad example. IMO it should not even exist. `EditAttributeBaseProps` and it's corresponding component is the result of trying to have one component do _everything_. And because it needs to do _everything_ you need to add boolean flags because you rarely have a parameter that should be used in _every_ scenario. A dropdown behaves very differently from a toggle-button, but since `EditAttributeBase` needs to handle _all_ input components, it needs boolean flags so it knows which scenario it is in and can behave accordingly. 

The reason that the new dropdown components still uses `EditAttributeBase` is because of the way Sofie saves data to the database from the frontend (which is handled by `EditAttributeBase`). It would have been at 10 times bigger refactor (if not more) to not be reliant on `EditAttributeBase`.